Use mappings for data model options and attributes (#800)
unicode-org/message-format-wg: spec/data-model/README.md

  type: "expression";
  arg: Literal;
  annotation?: FunctionAnnotation | UnsupportedAnnotation;
  attributes: Attributes;
}

interface VariableExpression {
  type: "expression";
  arg: VariableRef;
  annotation?: FunctionAnnotation | UnsupportedAnnotation;
  attributes: Attributes;
}

interface FunctionExpression {
  type: "expression";
  arg?: never;
  annotation: FunctionAnnotation;
  attributes: Attributes;
}

interface UnsupportedExpression {
  type: "expression";
  arg?: never;
  annotation: UnsupportedAnnotation;
  attributes: Attributes;
}

> type Attributes = Map<string, Literal | VariableRef | true>;

Reviewer: This changes the definition of attribute by adding `| true`. The syntax doesn't currently allow this?

Author: The `true` is meant to be used for representing value-less attributes like `@foo`, which the syntax does currently support.

A `FunctionAnnotation` represents a _function_ _annotation_.
The `name` does not include the `:` starting sigil.

> Each _option_ is represented by an `Options` key-value mapping.

Reviewer: Document the contents of `Options` in its own place and document `Options` here.

```suggestion
`Options` is a map containing the _expression_'s _options_.
```

Author: ```suggestion
`Options` is a key-value mapping containing options,
and is used to represent the _annotation_ and _markup_ _options_.
```

```ts
interface FunctionAnnotation {
  type: "function";
  name: string;
  options: Options;
}

> type Options = Map<string, Literal | VariableRef>;

Reviewer: Put documentation here for both `FunctionAnnotation` and Markup. (not using suggestion markdown because structural change would be needed)

> Options is a key-value map containing options.
> Implementations can use an ordered map if they want 
> to maintain the source ordering of options.
> Implementations might also canonicalize messages by ordering
> options into code point order.

each corresponding to _open_, _standalone_, and _close_ _markup_.
The `name` in these does not include the starting sigils `#` and `/` 
or the ending sigil `/`.

> The optional `options` for markup use the same key-value mapping as `FunctionAnnotation`.

Reviewer: "optional options" seems redundant? Consider making the text here exactly the same as in `FunctionAnnotation`

Author: ```suggestion
The `options` for markup use the same key-value mapping as `FunctionAnnotation`.
```
